metadata-io/src/main/java/com/linkedin/metadata/entity/EntityServiceImpl.java

```diff
@@ -408,23 +409,127 @@ public Map<Urn, List<EnvelopedAspect>> getLatestEnvelopedAspects(
       @Nonnull OperationContext opContext,
       @Nonnull Set<Urn> urns,
       @Nonnull Set<String> aspectNames,
-      boolean alwaysIncludeKeyAspect)
-      throws URISyntaxException {
+      boolean alwaysIncludeKeyAspect) {
 
-    final Set<EntityAspectIdentifier> dbKeys =
+    return getEnvelopedVersionedAspects(
+        opContext,
         urns.stream()
             .map(
                 urn ->
-                    aspectNames.stream()
-                        .map(
-                            aspectName ->
-                                new EntityAspectIdentifier(
-                                    urn.toString(), aspectName, ASPECT_LATEST_VERSION))
-                        .collect(Collectors.toList()))
-            .flatMap(List::stream)
+                    Map.entry(
+                        urn,
+                        aspectNames.stream()
+                            .map(aspectName -> Map.entry(aspectName, 0L))
+                            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)),
+        alwaysIncludeKeyAspect);
+  }
+
+  @Override
+  public Map<Urn, List<EnvelopedAspect>> getEnvelopedVersionedAspects(
+      @Nonnull OperationContext opContext,
+      @Nonnull Map<Urn, Map<String, Long>> urnAspectVersions,
```

Comment:
Might be too big of a refactor, but this feels like the structure should be:
`Map<Urn, Map<String, Set<Long>>>`

Where instead of doing this looping you just automatically add in version zero while it's constructed and then you can get them all in a single query rather than splitting it into two separate queries.

Reply:
Probably do-able, I'll take a look, ty!